torchgen/gen.py

```diff
@@ -1374,6 +1345,81 @@ def flatten_pre_group(
     )
 
 
+# Return native function declarations grouped by their namespaces.
+def get_native_function_declarations(
+    *,
+    grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
+    backend_indices: Dict[DispatchKey, BackendIndex],
+) -> List[str]:
+    declarations: List[str] = []
+    ns_grouped_kernels: Dict[str, List[str]] = defaultdict(list)
+    newline = "\n"
+    for f in grouped_native_functions:
+        for backend_idx in backend_indices.values():
+            backend_metadata = backend_idx.get_kernel(f)
+            namespace = (
+                backend_metadata.cpp_namespace
+                if backend_metadata
+                else DEFAULT_KERNEL_NAMESPACE
```

Comment:
So, reading this, it looks like we're implying full flexibility that for a given operator, it might have a different namespace for different backends.

I don't think that's actually true in practice, right? We don't even have a way to express that in the yaml syntax - from the yaml syntax, you'd expect that the cpp_namespace is a property of the `NativeFunction`, and not the backend.

The per-backend part comes mostly users of `gen_backend_stubs.py` and `gen_lazy_tensor.py`; if your a backend that uses one of those files to run codegen, you can specify a different namespace for your operator than the "default" specified from `native_functions.yaml` (so in one of your recent PR's, we slotted it into the `BackendIndex`). 

I wouldn't expect multiple namespaces to show up for a single operator for anything that ever invokes `gen.py` though (do you agree that's the case for mobile too?)

If that's the case, then maybe just assert that find all valid `BackendMetadata` objects for a given `NativeFunction`, and assert that their namespaces are all the same.

Reply:
Right now we don't have such use case, but wondering if we can do:

```
- func: cat.out(Tensor[] tensors, int dim=0, *, Tensor(a!) out) -> Tensor(a!)
  dispatch:
    CPU: cpu::cat_out
    CUDA: cuda::cat_out
```

Reply:
it feels a bit like unnecessary complexity to me haha (I don't think the codegen has logic to parse a namespace out of the kernel name today). But if that's important, we could probably do it.